src/dbnode/network/server/tchannelthrift/node/service.go

```diff
@@ -729,153 +728,291 @@ func (s *service) FetchTagged(tctx thrift.Context, req *rpc.FetchTaggedRequest)
 		)
 	}
 
-	result, err := s.fetchTagged(ctx, db, req)
+	result, err := s.fetchTagged(ctx, req)
 	if sampled && err != nil {
 		sp.LogFields(opentracinglog.Error(err))
 	}
 	sp.Finish()
 
+	if err != nil {
+		s.metrics.fetchTagged.ReportError(s.nowFn().Sub(callStart))
+	} else {
+		s.metrics.fetchTagged.ReportSuccess(s.nowFn().Sub(callStart))
+	}
+
 	return result, err
 }
 
-func (s *service) fetchTagged(ctx context.Context, db storage.Database, req *rpc.FetchTaggedRequest) (*rpc.FetchTaggedResult_, error) {
-	callStart := s.nowFn()
+func (s *service) fetchTagged(ctx context.Context, req *rpc.FetchTaggedRequest) (*rpc.FetchTaggedResult_, error) {
+	iter, err := s.FetchTaggedIter(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	response := &rpc.FetchTaggedResult_{
+		Elements:   make([]*rpc.FetchTaggedIDResult_, 0, iter.NumIDs()),
+		Exhaustive: iter.Exhaustive(),
+	}
+
+	for iter.Next(ctx) {
+		if iter.Err() != nil {
+			return nil, iter.Err()
+		}
+		cur := iter.Current()
+		tagBytes := make([]byte, 0)
+		tagBytes, err = cur.WriteTags(tagBytes)
+		if err != nil {
+			return nil, err
+		}
+		idResult := &rpc.FetchTaggedIDResult_{
+			ID:          cur.ID(),
+			NameSpace:   iter.Namespace().Bytes(),
+			EncodedTags: tagBytes,
+		}
+		response.Elements = append(response.Elements, idResult)
+		segIter := cur.SegmentsIter()
+		for segIter.Next(ctx) {
+			if segIter.Err() != nil {
```

Comment:
Same here would opt to more consistently check for the err after and have the `.Next(...)` call return false if there is an error to make it break.
```go
for segIter.Next(ctx) {
  // inner
}
if err := segIter.Err(); err != nil {
  return nil, err
}
```